src/mlpack/methods/ann/activation_functions/gelu_exact_function.hpp

```diff
@@ -0,0 +1,61 @@
+/**
+ * @file methods/ann/activation_functions/gelu_exact_function.hpp
+ *
+ * Definition and implementation of the exact Gaussian Error Linear Unit (GELU)
+ * function.
+ */
+#ifndef MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_GELU_EXACT_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_GELU_EXACT_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+
+/**
+ * The exact GELU function, defined by
+ *
+ * @f{eqnarray*}{
+ * f(x) = x * Phi(x) \\
+ * Phi(x) = 0.5 * (1 + erf(x / sqrt(2))) \\
+ * f'(x) = Phi(x) + x * phi(x) \\
+ * phi(x) = (1 / sqrt(2\pi)) * exp(-x^2 / 2)
+ * @f}
+ */
+class GELUExactFunction
+{
+ public:
+  //! Compute the exact GELU function for a single value.
+  static double Fn(const double x)
+  {
+    return 0.5 * x * (1.0 + std::erf(x / std::sqrt(2.0)));
+  }
+
+  //! Compute the exact GELU function for matrices/vectors.
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, OutputVecType& y)
+  {
+    y = 0.5 * x % (1.0 + arma::erf(x / std::sqrt(2.0)));
+  }
+
+  //! Compute the first derivative of the exact GELU function for a single value.
+  static double Deriv(const double x, const double /* y */)
+  {
+    const double phi = std::exp(-0.5 * x * x) / std::sqrt(2.0 * M_PI);
+    const double Phi = 0.5 * (1.0 + std::erf(x / std::sqrt(2.0)));
+    return Phi + x * phi;
```

Comment:
So, we actually have the result of `Fn(x)` already, as the parameter `y`.  Since `erf()` is expensive, better to just do one division instead.  So correct me if I'm wrong but I believe we could then simplify this expression to:

```
    return y / x + phi;
```

Reply:
I guess to prevent division by zero though we would need to either do `y / (x + 1e-15)` (and introduce some approximation) or do `return (x == 0.0) ? phi : (y / x + phi);`.